java/google/registry/rde/imports/XjcToDomainResourceConverter.java

```diff
@@ -181,14 +175,10 @@ public GracePeriod apply(XjcRgpStatusType gracePeriodStatus) {
   }
 
   /** Converts {@link XjcRdeDomain} to {@link DomainResource}. */
-  static DomainResource convertDomain(XjcRdeDomain domain) {
-    // First create history entry and autorenew billing event/poll message
-    // Autorenew billing event is required for creating AUTO_RENEW grace period
-    HistoryEntry historyEntry = createHistoryEntry(domain);
-    BillingEvent.Recurring autoRenewBillingEvent =
-        createAutoRenewBillingEvent(domain, historyEntry);
-    PollMessage.Autorenew pollMessage = createAutoRenewPollMessage(domain, historyEntry);
-    ofy().save().<ImmutableObject>entities(historyEntry, autoRenewBillingEvent, pollMessage);
+  static DomainResource convertDomain(
+      XjcRdeDomain domain,
+      BillingEvent.Recurring autoRenewBillingEvent,
+      PollMessage.Autorenew autoRenewPollMessage) {
     GracePeriodConverter gracePeriodConverter =
         new GracePeriodConverter(domain, Key.create(autoRenewBillingEvent));
```

Comment:
This comment really applies more specifically to above where GracePeriodConverter.apply() calls GracePeriod.createForRecurring(), but GitHub apparently doesn't let you comment outside diff ranges.

I think using createForRecurring() in this context is incorrect.  It's being called with the Recurring billing event that we're passing in to this function, which we just newly created via createAutoRenewBillingEventForDomainImport().  The problem is, that Recurring event has an eventTime set to the domain's expiration time, which must be in the future relative to the import time.  Any given Recurring event is only expanded (e.g. by the ExpandRecurringBillingEventsAction) within the time range defined by [eventTime, recurrenceEndTime).  So the effect is that the grace period, if invoked, will trigger a cancellation that doesn't correspond to any actual billing event -  the cancellation will be for an autorenew that happened in the past, but the recurring event we're using doesn't start recurring until the future.

I'm curious more broadly about what you're expecting to do about billing for grace periods that span the migration from one TLD to another.  None of the other grace periods in GracePeriodConverter are being given billing events, so they won't generate any cancellations if the domain is deleted while still in the grace period.  Are you thinking that you'd just do some kind of one-time manual process to inspect for any domains deleted shortly after the migration, determine what grace periods they would have triggered, and issue refunds appropriately?

At any rate, while it's possible to hack around the autorenew issue (e.g. by setting the eventTime of the autorenew to the creationTime of the domain plus 1 year, which still may not be historically accurate but at least fixes the dangling cancellation problem), it seems like any reconciliation process you have would be easier to manage if it applied to all types of grace periods, rather than trying to special-case autorenews to do something cleverer.